payjoin/src/receive/mod.rs

```diff
@@ -382,62 +383,61 @@ impl WantsOutputs {
-                    )
-                    .into()),
-                _ => log::info!("Receiver is augmenting outputs"),
-            }
-        }
         let mut outputs = vec![];
         let mut replacement_outputs = replacement_outputs.clone();
         let mut rng = rand::thread_rng();
-        // Replace the existing receiver outputs
-        for (i, output) in self.original_psbt.unsigned_tx.output.iter().enumerate() {
+        // Substitute the existing receiver outputs, keeping the sender/receiver output ordering
+        for (i, original_output) in self.original_psbt.unsigned_tx.output.iter().enumerate() {
             if self.owned_vouts.contains(&i) {
-                // Receiver output: substitute with a provided output picked randomly
+                // Receiver output: substitute in-place a provided replacement output
                 if replacement_outputs.is_empty() {
                     return Err(InternalOutputSubstitutionError::NotEnoughOutputs.into());
                 }
-                let index = rng.gen_range(0..replacement_outputs.len());
-                let txo = replacement_outputs.swap_remove(index);
-                if *drain_script == txo.script_pubkey {
-                    change_vout = Some(i);
+                match replacement_outputs
+                    .iter()
+                    .position(|txo| txo.script_pubkey == original_output.script_pubkey)
+                {
+                    // Select an output with the same address if one was provided
+                    Some(pos) => {
+                        let txo = replacement_outputs.remove(pos);
+                        if self.params.disable_output_substitution
+                            && txo.value < original_output.value
+                        {
+                            return Err(
+                                InternalOutputSubstitutionError::OutputSubstitutionDisabled(
+                                    "Decreasing the receiver output value is not allowed",
+                                )
+                                .into(),
+                            );
+                        }
+                        outputs.push(txo);
+                    }
+                    // Otherwise randomly select one of the replacement outputs
+                    None => {
+                        if self.params.disable_output_substitution {
+                            return Err(
+                                InternalOutputSubstitutionError::OutputSubstitutionDisabled(
+                                    "Changing the receiver output script pubkey is not allowed",
+                                )
+                                .into(),
+                            );
+                        }
+                        let index = rng.gen_range(0..replacement_outputs.len());
+                        let txo = replacement_outputs.swap_remove(index);
```

Comment:
I assume swap_remove is used here for performance because the order of replacement outputs is assumed to be unpreserved (we should document that).

Could that also apply to removal in the Some(pos) condition above, or is there a reason `remove` was chosen there instead? I'm had to look up docs for swap_remove I haven't seen it before but it does make sense to me now.

```
Some(pos) => {
        let txo = replacement_outputs.remove(pos);?
```

Reply:
Good point, I had selected because of the performance warning in the `remove()` docstring but didn't apply it evenly here. I'll update that other one.